controllers/users.js

```diff
@@ -1,33 +1,95 @@
 const mongoose = require('mongoose');
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
 
 const User = require('../models/user.js');
 
+const BodyDataValidError = require('../errors/bodyDataValidError');
+const FindIdUserError = require('../errors/findIdUserError');
+const IdValidError = require('../errors/idValidError');
+
+const serverError = require('../errors/serverError');
+
 const showOneUser = (req, res) => {
   const validId = mongoose.Types.ObjectId.isValid(req.params._id);
 
-
   if (validId) {
-    User.findById(req.params._id)
-      .then((user) => (user ? res.send({ data: user }) : res.status(404).send({ message: 'пользователь не найден' })))
-      .catch(() => res.status(500).send({ message: 'Произошла ошибка' }));
-  } else {
-    res.status(400).send({ message: 'Невалидный id' });
+    return User.findById(req.params._id)
+      .then((user) => (user ? res.send({ data: user }) : Promise.reject(new FindIdUserError('Пользователь не найден'))))
+      .catch((err) => res
+        .status(err.statusCode || 500)
+        .send({
+          message: err.message || serverError,
+        }));
   }
+  const { message, statusCode } = new IdValidError('Невалидный id');
+
+  return res
+    .status(statusCode)
+    .send({
+      message,
+    });
 };
 
 const showAllUsers = (req, res) => {
   User.find({})
     .then((users) => res.send({ data: users }))
-    .catch(() => res.status(500).send({ message: 'Произошла ошибка' }));
+    .catch((err) => res.status(500).send(err.message));
 };
 
-
 const postUser = (req, res) => {
-  const { name, about, avatar } = req.body;
+  const {
+    name, about, avatar, email, password,
+  } = req.body;
+
+  return User.validLogin(email, password)
+    .then(() => User.existingUser(email)
+      .then(() => bcrypt.hash(password, 10)
+        .then((hash) => User.create({
+          name, about, avatar, email, password: hash,
+        }))
+        .then((user) => res.status(201).send({
+          data: user.omitPrivate(),
+        }))))
+    .catch((err) => {
+      if (err._message === 'user validation failed') {
+        const { statusCode } = new BodyDataValidError();
 
-  User.create({ name, about, avatar })
-    .then((user) => res.send({ data: user }))
-    .catch(() => res.status(500).send({ message: 'Произошла ошибка' }));
+        return res
+          .status(statusCode)
+          .send({
+            message: err.message,
+          });
+      }
+
+      return res
+        .status(err.statusCode || 500)
+        .send({
+          message: err.message || serverError,
+        });
```

Comment:
Можно лучше: унифицировать текст ответа для 500.
Клиенту врядли удастся согласно реальному тексту ошибки что-то исправить, чтобы корректно заработал контроллер. Поэтому нет смысла его отправлять. Рекомендую для 500 отправлять просто какой-то текст:
```suggestion
      const statusCode = err.statusCode || 500;
      return res
        .status(statusCode)
        .send({
          message:statusCode === 500 ? 'Ошибка на сервере' : err.message,
        });
```